File: app/src/main/java/com/example/desiregallery/network/serializers/PostListDeserializer.java
package com.example.desiregallery.network.serializers;

import com.example.desiregallery.models.Post;
import com.google.gson.*;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class PostListDeserializer implements JsonDeserializer<List<Post>> {
    @Override
    public List<Post> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        List<Post> posts = new ArrayList<Post>();
        JsonArray documents = json.getAsJsonObject().get("documents").getAsJsonArray();

        for (JsonElement el : documents) {
            Post post = context.deserialize(el, Post.class);
            posts.add(post);
        }
        return posts;
    }
}

File: app/src/main/java/com/example/desiregallery/network/serializers/PostListDeserializer.kt
package com.example.desiregallery.network.serializers

import com.example.desiregallery.models.Post
import com.example.desiregallery.models.User
import com.example.desiregallery.network.getUsersByNames
import com.google.gson.*

import java.lang.reflect.Type

class PostListDeserializer : JsonDeserializer<List<Post>> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): List<Post>  {
        val documents = json.asJsonArray.filter { it.asJsonObject.has("document") }
        val posts: List<Post> = documents
            .map { context.deserialize(it.asJsonObject.get("document"), Post::class.java) as Post }

        val authors: Set<String> = LinkedHashSet(posts.map { post -> post.author.login })
        val users = getUsersByNames(authors)
        posts.map { post ->
            val authorName = post.author.login
            post.author = users.find { user -> user.login == authorName } as User
        }
        return posts
    }
}
